File: cpp/ql/test/query-tests/Security/CWE/CWE-020/ExternalAPITaintStepExample.cpp
typedef unsigned long size_t;
typedef size_t FILE;

char *strcat(char *s1, const char *s2);
char *fgets(char *s, int n, FILE *stream);
char *fputs(const char *s, FILE *stream);

void do_get(FILE* request, FILE* response) {
  char user_id[1024];
  fgets(user_id, 1024, request);

  char buffer[1024];
  strcat(buffer, "SELECT * FROM user WHERE user_id='");
  strcat(buffer, user_id);
  strcat(buffer, "'");

  fputs(buffer, response);
}

File: java/ql/test/query-tests/security/CWE-020/ExternalAPITaintStepExample.java
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.ServletException;
import java.io.IOException;

public class ExternalAPITaintStepExample extends HttpServlet {
	protected void doGet(HttpServletRequest request, HttpServletResponse response)
	throws ServletException, IOException {

		StringBuilder sqlQueryBuilder = new StringBuilder();
		sqlQueryBuilder.append("SELECT * FROM user WHERE user_id='");
		// BAD: a request parameter is concatenated directly into a SQL query
		sqlQueryBuilder.append(request.getParameter("user_id"));
		sqlQueryBuilder.append("'");

		// ...
	}
}
